src/sdk/namespace_cache.js

```diff
@@ -124,27 +214,193 @@ class NamespaceCache {
         }
     }
 
-    async read_object_stream(params, object_sdk) {
+    // If object is not cached,
+    //   - it creates partial object and uploads part to cache if range read is performed on hub
+    //   - it uploads entire object to cache if entire read is performed on hub.
+    //
+    // If entire or partial object is cached,
+    //   - if entire read is performed on hub and cache does not have all the data,
+    //     it uploads entire object to cache .
+    //   - otherwise, it reads data from cache by providing missing_part_getter to
+    //     cache's read_object_stream.
+    async _read_object_stream(params, object_sdk) {
+        dbg.log0('NamespaceCache._read_object_stream', {params: params});
+
+        params.max_cache_obj_size = await this._get_max_cache_obj_size_bytes(params, object_sdk);
+        const range_hub_read = this._range_read_hub_check(params);
+        if (!params.object_md.should_read_from_cache) {
+            // Object not in cache
+            if (range_hub_read && params.read_size <= params.max_cache_obj_size) {
+                const create_params = _.pick(params,
+                    'bucket',
+                    'key',
+                    'content_type',
+                    'size',
+                    'etag',
+                    'xattr',
+                    'partial_object'
+                );
+                create_params.size = params.object_md.size;
+                create_params.etag = params.object_md.etag;
+                create_params.xattr = params.object_md.xattr;
+                create_params.partial_object = true;
+                // Create partial object md
+                const create_reply = await object_sdk.rpc_client.object.create_object_upload(create_params);
+                params.object_md.obj_id = create_reply.obj_id;
+
+                dbg.log0('NamespaceCache._read_object_stream: partial object created:', create_reply.obj_id);
+            }
+
+            return this._read_hub_object_stream(params, object_sdk,
+                range_hub_read ? { start: params.start, end: params.end } : undefined);
+        }
+
+        if (!range_hub_read && params.object_md.upload_size < params.object_md.size) {
+            // Cache does not have all the data. Perform entire hub read
+            return this._read_hub_object_stream(params, object_sdk);
+        }
```

Comment:
either remove this, or insert into the missing_part_getter logic later to use a full object read rather than a ranged read.

Reply:
ok